src/prep.py

```diff
@@ -0,0 +1,331 @@
+    print(
+        f"You should not run {__file__} as a standalone script.\n"
+        f"Use `$ python3 reduce.py {Path(__file__).stem}` instead!",
+        file=sys.stderr,
+    )
+    sys.exit(1)
+
+
+DEBUG = False
+ic.disable()
+
+
+def make_dot_i_file(
+    build_cmd: Command,
+    target: Path,
+    path_to_linux: Path,
+    output_dir: Path,
+    force_rm_existing_target: bool,
+) -> str:
+    """
+    Make the preprocessed .i file which can then be compiled sans build system
+    """
+
+    if (abs_target := path_to_linux / target).exists():
+        double_check_removal_with_user(
+            file=abs_target, force_rm=force_rm_existing_target
+        )
+
+    info(f"Making {target}...")
+    make_output = subprocess.run(
+        [*build_cmd, target], check=False, capture_output=True, cwd=path_to_linux
+    )
+    ic(make_output)
+
+    if make_output.returncode == 0:
+        info(f"Moving generated {target} file to {output_dir}")
+        shutil.move(path_to_linux / target, output_dir / target.name)
+        success(f"Successfully generated {target.name}")
+
+    return make_output.stdout.decode("utf-8")
+
+
+def get_compiler_invocation(target: Path, make_stdout: str) -> Command:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = f".*-o {target}.*"
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches!\n\
+            (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    main_invocation.insert(-1, "-c")
+    return clean_compiler_invocation(main_invocation)
+
+
+def clean_compiler_invocation(cc_invocation: Command) -> Command:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+
+    to_remove = ("-I", "-D", "-Wp", "-include", "-Werror", "./", "-U", "-E")
+    rules = [(lambda rule: lambda flag: rule not in flag)(rule) for rule in to_remove]
+
+    cleaned = cc_invocation.copy()
+    for rule in rules:
+        cleaned = filter(rule, cleaned)
+
+    cleaned = list(cleaned)  # eagerly convert iterator to list
+
+    ic(cleaned)
+
+    return cleaned
+
+
+def write_test_script(
+    *,
+    cc_invocation: Command,
+    target: Path,
+    output_dir: Path,
+    force_rm: bool,
+    uses_clang: bool,
+    go_fast: bool,
+) -> None:
+    """
+    Create a test.sh script at --output directory
+    """
+    script_path = output_dir / Path("test.sh")
+    info(f"Generating {script_path}")
+    if script_path.exists():
+        double_check_removal_with_user(file=script_path, force_rm=force_rm)
+
+    output_target_idx = cc_invocation.index("-o") + 1
+    cc_invocation[output_target_idx] = target.name
+
+    cc_flags = "\n".join(
+        str(x) for x in cc_invocation[1:-4]
+    )  # exclude [clang/gcc] call and -o -c flags
+
+    target_flags = " ".join(
+        str(x) for x in cc_invocation[-4:]
```

Comment:
what is this slicing trying to do? Not clear from the comments

Reply:
turn

`clang -Wall -Wsomething -Wotherthing -xyz --target=thermometer -o target.out -c target.i`

into `-Wall -Wsomething -Wotherthing -xyz --target=thermometer`

essentially ditching the `clang` and `-o target.out -c target.i` so I can write them to a flags.txt file. Our flags.txt file need not modify our compiler invocation nor our target i/o.

Reply:
Any ideas on a more elegant approach than magic-number slicing?

Reply:
Assuming that `-o foo.o` and `-c input.i` are the last 4 positional command line args is brittle.  It's probably the case for all code in the kernel today, but I don't know that is true or always will be.

Explicitly filtering out `-o <and what follows -o>` is more self documenting than slicing by magic values.  Same for `-c` and the input filename.

This is not the case with the initial arg, since `clang` can't appear like `-Wall -Wsomething clang -Wotherthing -xyz --target=thermometer -o target.out -c target.i`

Reply:
Fixed. Thanks.